haystack/dataclasses/tool.py

```diff
@@ -0,0 +1,18 @@
+from typing import Any, Callable, Dict
+
+
+class Tool:
+    def __init__(self, name: str, description: str, parameters: Dict[str, Any], function: Callable):
```

Comment:
I'd make the `Tool` class an `ABC` that exposes properties for the basic fields:
- `name: str`
- `description: str`
- `params: Dict[str, ParamInfo]` (`ParamInfo` would contain generic fields for docstrings, type hints, defaults, etc)

Then, we can create impl.-specific tools as subclasses, e.g: `FunctionTool`, etc. Take a look at how the `Secret` class and its subclasses are implemented for the general pattern.

Reply:
I’d like to clarify my proposal.

The LLM (OpenAI/Anthropic/Ollama...) only needs three pieces of information to prepare a call for the tool: `name`, `description`, and `parameters` (a JSON schema).

To make things easier for users, we can add some utilities to automatically convert various entities into a tool. However, the core requirements for the LLM remain unchanged.

So I would advocate for keeping this class very simple.
We can then add, for example, some class methods to convert different entities into tools.
An example:
```python
@classmethod
from_function(cls, function, ...)

  # some automatical extraction of data
  ...

  return cls(name=extracted_name,
    description=extracted_description,
    parameters=extracted_parameters,
    function=function)
```			  

If it is meaningful and/or feasible, we can also add `from_component` and `from_pipeline` methods.

WDYT?

Reply:
If you feel confident that those three fields are enough to handle tool usage for most/all generators and we won't need extra tool-specific logic, go ahead with your proposal. 

Re. parameters - Is there a standard schema used by all providers? If so, I'd rather have an explicit dataclass that defines it than `Any`.